pkg/ruler/evaluator_remote.go

```diff
@@ -0,0 +1,263 @@
+
+	queryEndpointPath = "/loki/api/v1/query"
+
+	mimeTypeFormPost = "application/x-www-form-urlencoded"
+)
+
+const EvalModeRemote = "remote"
+
+var userAgent = fmt.Sprintf("loki-ruler/%s", build.Version)
+
+type RemoteEvaluator struct {
+	rq     *remoteQuerier
+	logger log.Logger
+}
+
+func NewRemoteEvaluator(cfg *EvaluationConfig, logger log.Logger) (*RemoteEvaluator, error) {
+	qfClient, err := dialQueryFrontend(cfg.QueryFrontend)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial query frontend for remote rule evaluation: %w", err)
+	}
+
+	return &RemoteEvaluator{
+		rq:     newRemoteQuerier(qfClient, logger, WithOrgIDMiddleware),
+		logger: logger,
+	}, nil
+}
+
+func (r *RemoteEvaluator) Eval(ctx context.Context, qs string, now time.Time) (*logqlmodel.Result, error) {
+	res, err := r.rq.Query(ctx, qs, now)
+	if err != nil {
+		return nil, fmt.Errorf("failed to perform remote evaluation of query %q: %w", qs, err)
+	}
+
+	return res, err
+}
+
+// dialQueryFrontend creates and initializes a new httpgrpc.HTTPClient taking a QueryFrontendConfig configuration.
+func dialQueryFrontend(cfg QueryFrontendConfig) (httpgrpc.HTTPClient, error) {
+	tlsDialOptions, err := cfg.TLS.GetGRPCDialOptions(cfg.TLSEnabled)
+	if err != nil {
+		return nil, err
+	}
+	dialOptions := append(
+		[]grpc.DialOption{
+			grpc.WithKeepaliveParams(
+				keepalive.ClientParameters{
+					Time:                keepAlive,
+					Timeout:             keepAliveTimeout,
+					PermitWithoutStream: true,
+				},
+			),
+			grpc.WithUnaryInterceptor(
+				grpc_middleware.ChainUnaryClient(
+					otgrpc.OpenTracingClientInterceptor(opentracing.GlobalTracer()),
+					middleware.ClientUserHeaderInterceptor,
+				),
+			),
+			grpc.WithDefaultServiceConfig(serviceConfig),
+		},
+		tlsDialOptions...,
+	)
+
+	conn, err := grpc.Dial(cfg.Address, dialOptions...)
+	if err != nil {
+		return nil, err
+	}
+	return httpgrpc.NewHTTPClient(conn), nil
+}
+
+// Middleware provides a mechanism to inspect outgoing remote querier requests.
+type Middleware func(ctx context.Context, req *httpgrpc.HTTPRequest) error
+
+// remoteQuerier executes read operations against a httpgrpc.HTTPClient.
+type remoteQuerier struct {
+	client      httpgrpc.HTTPClient
+	middlewares []Middleware
+	logger      log.Logger
+}
+
+// newRemoteQuerier creates and initializes a new remoteQuerier instance.
+func newRemoteQuerier(
+	client httpgrpc.HTTPClient,
+	logger log.Logger,
+	middlewares ...Middleware,
+) *remoteQuerier {
+	return &remoteQuerier{
+		client:      client,
+		middlewares: middlewares,
+		logger:      logger,
+	}
+}
+
+// Query performs a query for the given time.
+func (q *remoteQuerier) Query(ctx context.Context, qs string, t time.Time) (*logqlmodel.Result, error) {
+	logger, ctx := spanlogger.NewWithLogger(ctx, q.logger, "ruler.remoteEvaluation.Query")
+	defer logger.Span.Finish()
+
+	res, err := q.query(ctx, qs, t, logger)
+	if err != nil {
+		return nil, err
+	}
+	return res, err
```

Comment:
nit: should be identical to:

```suggestion
	return q.query(ctx, qs, t, logger)
```

Reply:
You're right, I think I had some error wrapping there before but removed it.